### ISSUE TITLE
Clarity and control on a JsonDocument lifetime

### ISSUE THREAD

**Issue by bartonjs** [dotnet/corefx#36152]
There are situations where a library may receive a `JsonElement` and want to hold this value for longer than the method lifetime.  In the current model, assuming the caller of JsonDocument.Parse did so in a `using` statement, the JsonElement will get invalidated.

Similarly, but differently, a user of JsonDocument may find that they want to reclaim/reuse the original payload memory only after seeing the contents of the document.

# Simplified Proposal
```C#
public partial struct JsonElement
{
    // When true, indicates that this object will not ever degrade into an ObjectDisposedException
    public bool IsPersistable { get; }

    // Makes a copy (when required) of this value which can be persisted (separated from the
    // JsonDocument Dispose semantics.
    //
    // Is no-op when IsPersistable is true, and this is equivalent to a JsonDocument.RootElement.
    public JsonElement Clone();
}
```

## Original Proposal

```C#
public partial class JsonDocument
{
    /// <summary>
    ///   Whether or not this JsonDocument is detached from memory provided during a call to Parse.
    /// </summary>
    public bool IsDetached { get; }

    /// <summary>
    ///   Whether or not this JsonDocument should be disposed when no longer needed.
    /// </summary>
    /// <remarks>
    ///   The default behavior for JsonDocument is to utilize pooled arrays for its data, and
    ///   <see cref="Dispose"/> returns those arrays to their respective pools. A call to
    ///   <see cref="Detach"/> can indicate whether the utilized memory comes from the array
    ///   pools, or new arrays held by this type. When new arrays are used, Dispose has no effect
    ///   and therefore no longer needs to be called.
    /// </remarks>
    public bool IsDisposable { get; }

    /// <summary>
    ///   Get a JsonDocument representing the same contents as the current object,
    ///   but which does not depend on the data provided to
    ///   <see cref="Parse(ReadOnlyMemory{byte},JsonReaderOptions)"/> (or another overload)
    ///   remaining unchanged.
    /// </summary>
    /// <param name="useArrayPools">
    ///   <see langword="true"/> to use pooled arrays where possible, <see langword="false"/> to
    ///   use newly created arrays for simpler lifetime management.
    ///   (Defaults to <see langword="false"/>.)
    /// </param>
    /// <remarks>
    ///   <para>
    ///     This method returns <see langword="this"/> when doing so is only detectable via
    ///     a reference equality test. If the current instance is already detached (
    ///     <see cref="IsDetached"/> == <see langword="true"/>) and has GC lifetime semantics
    ///     instead of IDisposable lifetime (<see cref="IsDisposable"/> == <see langword="false"/>)
    ///     then no functional side effects can be observed by multiple callers acting on the
    ///     same instance.
    ///   </para>
    ///   <para>
    ///     When invoking this method with <paramref name="useArrayPools"/> == <see langword="true"/>,
    ///     the caller is responsible for managing the lifetime of the returned object (as an
    ///     <see cref="IDisposable"/>). When <paramref name="useArrayPools"/> == <see langword="false"/>,
    ///     <see cref="Dispose"/> has no effect and therefore the caller has no lifetime management
    ///     responsibilities.
    ///   </para>
    /// </remarks>
    /// <returns>
    ///   A JsonDocument instance representing the same contents, but with memory independent
    ///   of what was provided to <see cref="Parse(ReadOnlyMemory{byte},JsonReaderOptions)"/> (or
    ///   another overload).
    /// </returns>
    public JsonDocument Detach(bool useArrayPools=false)
}

public partial struct JsonElement
{
    // Indicates that the JsonElement can be held long term.  Therefore it keys off of
    // JsonDocument.IsDIsposable, not IsDetached.
    public bool IsDetached { get; }

    // If IsDetached and this is its documents RootElement, return this.
    // Otherwise returns a copy (using strictly less memory) which is GC-tracked.
    public JsonElement Detach();
}
```

Bonus extra:

```C#
public partial class JsonDocument
{
    // This allows for an ArrayPool/Disposable version of JsonElement.Detach
    public static JsonDocument DetachAt(JsonElement element, bool useArrayPools=false);

}

// alternatively:

public partial struct JsonElement
{
    public JsonDocument DetachDocument(bool useArrayPools=false);
}
```

JsonDocument.Parse's overloads will (be honest and) leak some of their implementation details via `IsDetached`: The operations which copy the input into a sequential UTF-8 buffer (and thus no longer depend on the input being pristine) will report `IsDetached` as `true`.

The different states (equivalence class):

* `JsonDocument.Parse(ReadOnlyMemory<byte>)`
  * Metadata is stored in ArrayPool-based arrays, GetInt32() (etc) go all the way back to this parameter's memory.
    * Dispose puts the arrays back, and invalidates the JsonDocument and any JsonElements from the JsonDocument.
  * IsDetached == false
  * IsDisposable == true
* doc.Detach(true)
  * The JSON payload is copied into an ArrayPool-based array, and the metadata is in ArrayPool-based arrays.  No dependence remains on the original memory.
    * Dispose puts all of the arrays back, invalidates the JsonDocument and any JsonElements from the JsonDocument.
  * IsDetached == true
  * IsDisposable == true
* doc.Detach(false)
  * The JSON payload is copied into `new byte[payloadLength]` and the metadata is copied into `new byte[mdLength]`; making the object a classic GC-tracked object lifetime.
    * Dispose is a no-op.
  * IsDetached == true
  * IsDisposable == false

The fourth state (IsDisposable == false, IsDetached == false) is not possible.

## Naming

One possibility for JsonDocument is to replace the two booleans with one enumeration; but it doesn't have really good names, either.

```C#
enum JsonDocumentLifetime
{
    PooledExternalReference,
    PooledOnly,
    GC,
}
```

`IsDisposable` is an okay description of what effects are visible, but seems to be lacking in the naming department.  Other considered options `IsFullyDetached`/`IsDetachedFully`, `RequiresDispose`/`NeedsDispose`/`NeedsDisposing`.  Definitely open to ideas.

It's also a bit unfortunate that JsonDocument and JsonElement both have IsDetached, but they mean different things.

JsonDocument: `UsingPooledMemory` and `DetachedFromInput`?

**Comment by terrajobst** [comment]
Let's remove `IsPersistable` for now and make `Clone()` a no-op. This way, callers can always call `Clone()` before storing it.
